### PR TITLE
Add sort option:Only this folder.

### PR DESCRIPTION
**Pull request by ningshu**
Fixes [#528](https://github.com/TeamAmaze/AmazeFileManager/issues/528)

### CODE REVIEW

File: app/src/main/java/com/amaze/filemanager/database/models/Sort.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+/**
+ * Created by Ning on 5/28/2018.
+ */
+public class Sort {
+    public static final int SORT_TYPE_NAME = 0;
+    public static final int SORT_TYPE_LAST_MODIFIED = SORT_TYPE_NAME + 1;
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/database/models/Sort.java]: Please hardcode these all the way (1, 2, 3, etc.).

File: app/src/main/java/com/amaze/filemanager/database/SortHandler.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+                + " INTEGER PRIMARY KEY,"
+                + COLUMN_SORT_PATH + " TEXT," + COLUMN_SORT_TYPE + " TEXT" + ")";
+
+        db.execSQL(CREATE_TABLE_SORT);
+    }
+
+    @Override
+    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
+
+        db.execSQL("DROP TABLE IF EXISTS " + TABLE_SORT);
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/database/SortHandler.java]: Remove.

File: app/src/main/java/com/amaze/filemanager/database/SortHandler.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        this.context = context;
+    }
+
+    @Override
+    public void onCreate(SQLiteDatabase db) {
+
+        String CREATE_TABLE_SORT = "CREATE TABLE " + TABLE_SORT + "("
+                + COLUMN_SORT_ID
+                + " INTEGER PRIMARY KEY,"
+                + COLUMN_SORT_PATH + " TEXT," + COLUMN_SORT_TYPE + " TEXT" + ")";
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/database/SortHandler.java]: Why is `COLUMN_SORT_TYPE` text? Shouldn't it be integer?

File: app/src/main/java/com/amaze/filemanager/database/SortHandler.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
+
+        db.execSQL("DROP TABLE IF EXISTS " + TABLE_SORT);
+        onCreate(db);
+    }
+
+    public void addEntry(Sort sort) {
+
+        ContentValues contentValues = new ContentValues();
+        //contentValues.put(COLUMN_SORT_ID, encryptedEntry.getId());
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/database/SortHandler.java]: Comment, add TODO or delete.

**Comment by EmmanuelMess** [comment]
3 tables will be deleted after this update:
```
    @Override
    public void onUpgrade(SQLiteDatabase sqLiteDatabase, int i, int i2) {
       sqLiteDatabase.execSQL("DROP TABLE IF EXISTS " + TABLE_TAB);
        sqLiteDatabase.execSQL("DROP TABLE IF EXISTS " + TABLE_ENCRYPTED);
        sqLiteDatabase.execSQL("DROP TABLE IF EXISTS " + CloudHandler.TABLE_CLOUD_PERSIST);
        //...
    }
```

File: app/src/main/java/com/amaze/filemanager/database/models/Sort.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+            SORT_TYPE_NAME_DESC,
+            SORT_TYPE_LAST_MODIFIED_DESC,
+            SORT_TYPE_SIZE_DESC,
+            SORT_TYPE_TYPE_DESC
+    })
+    @Retention(RetentionPolicy.SOURCE)
+    public @interface TYPE {
+    }
+
+    private int _id;
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/database/models/Sort.java]: Should be `public final id`.

File: app/src/main/java/com/amaze/filemanager/database/models/Sort.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    public String path;
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/database/models/Sort.java]: `public final` please.

File: app/src/main/java/com/amaze/filemanager/database/models/Sort.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    @TYPE
+    public int sortType;
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/database/models/Sort.java]: `public final` please.

File: app/src/main/java/com/amaze/filemanager/database/models/Sort.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+    public Sort() {
+    }
+
+    public Sort(String path, @TYPE int sortType) {
+        this.path = path;
+        this.sortType = sortType;
+    }
+
+    public int getId() {
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/database/models/Sort.java]: Make class immutable and remove getters.

**Comment by EmmanuelMess** [comment]
First sorting by size "only this folder" and the sorting by name (not only this folder) doesn't sort by name, remember that sort by x "only this folder" preference should be revertible.

File: app/src/main/java/com/amaze/filemanager/database/SortHandler.java [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+            sqLiteDatabase.delete(TABLE_SORT, COLUMN_SORT_PATH + " = ?", new String[]{path});
+        } catch (NumberFormatException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public void updateEntry(Sort oldSort, Sort newSort) {
+        SQLiteDatabase sqLiteDatabase = getWritableDatabase();
+        ContentValues contentValues = new ContentValues();
+        contentValues.put(COLUMN_SORT_ID, newSort.getId());
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/database/SortHandler.java]: This is redundant. You later have a where clause with `COLUMN_SORT_ID + " = ?"` where ? is `oldSort.getId() + ""`.

**Comment by VishalNehra** [comment]
Getting force close, can't test
Logcat:
````
05-28 21:00:07.185 11507-11507/com.amaze.filemanager E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.amaze.filemanager, PID: 11507
    android.database.sqlite.SQLiteException: no such table: sort (code 1): , while compiling: Select * FROM sort WHERE path= "/storage/emulated/0"
        at android.database.sqlite.SQLiteConnection.nativePrepareStatement(Native Method)
        at android.database.sqlite.SQLiteConnection.acquirePreparedStatement(SQLiteConnection.java:890)
        at android.database.sqlite.SQLiteConnection.prepare(SQLiteConnection.java:501)
        at android.database.sqlite.SQLiteSession.prepare(SQLiteSession.java:588)
        at android.database.sqlite.SQLiteProgram.<init>(SQLiteProgram.java:58)
        at android.database.sqlite.SQLiteQuery.<init>(SQLiteQuery.java:37)
        at android.database.sqlite.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:46)
        at android.database.sqlite.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:1392)
        at android.database.sqlite.SQLiteDatabase.rawQuery(SQLiteDatabase.java:1331)
        at com.amaze.filemanager.database.SortHandler.findEntry(SortHandler.java:106)
        at com.amaze.filemanager.database.SortHandler.getSortType(SortHandler.java:34)
        at com.amaze.filemanager.fragments.MainFragment.getSortModes(MainFragment.java:1398)
        at com.amaze.filemanager.fragments.MainFragment.onActivityCreated(MainFragment.java:266)
        at android.support.v4.app.Fragment.performActivityCreated(Fragment.java:2270)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1442)
        at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1750)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1819)
        at android.support.v4.app.BackStackRecord.executeOps(BackStackRecord.java:797)
        at android.support.v4.app.FragmentManagerImpl.executeOps(FragmentManager.java:2590)
        at android.support.v4.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2377)
        at android.support.v4.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManager.java:2332)
        at android.support.v4.app.FragmentManagerImpl.execSingleAction(FragmentManager.java:2209)
        at android.support.v4.app.BackStackRecord.commitNowAllowingStateLoss(BackStackRecord.java:649)
        at android.support.v4.app.FragmentStatePagerAdapter.finishUpdate(FragmentStatePagerAdapter.java:167)
        at android.support.v4.view.ViewPager.populate(ViewPager.java:1238)
        at android.support.v4.view.ViewPager.populate(ViewPager.java:1086)
        at android.support.v4.view.ViewPager.onMeasure(ViewPager.java:1616)
        at android.view.View.measure(View.java:22091)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6602)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:185)
        at android.view.View.measure(View.java:22091)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6602)
        at android.support.design.widget.CoordinatorLayout.onMeasureChild(CoordinatorLayout.java:719)
        at android.support.design.widget.HeaderScrollingViewBehavior.onMeasureChild(HeaderScrollingViewBehavior.java:91)
        at android.support.design.widget.AppBarLayout$ScrollingViewBehavior.onMeasureChild(AppBarLayout.java:1361)
        at android.support.design.widget.CoordinatorLayout.onMeasure(CoordinatorLayout.java:789)
        at android.view.View.measure(View.java:22091)
        at android.support.v4.widget.DrawerLayout.onMeasure(DrawerLayout.java:1060)
        at android.view.View.measure(View.java:22091)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6602)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:185)
        at android.support.v7.widget.ContentFrameLayout.onMeasure(ContentFrameLayout.java:139)
        at android.view.View.measure(View.java:22091)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6602)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:185)
        at android.view.View.measure(View.java:22091)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6602)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:185)
        at android.view.View.measure(View.java:22091)
````

**Comment by EmmanuelMess** [comment]
Call `getReadableDatabase()` the get onUpgrade to run.

**Comment by ningshu** [comment]
Thanks for your suggestions.
I will do more test before this PR.

**Comment by ningshu** [comment]
The creation of the new database table 'Sort' need to be compatible with the existing database upgrade mechanism.
Some member variables in file MainFragment.java ( sortby, dsort, asc ) need to refactor.

Close this defective PR.

**Comment by EmmanuelMess** [comment]
You can just update this branch until it is ready. Then we can reopen the PR.

**Pull request closed without merging**